File: RegexEngine.java
import java.util.Scanner;

public class RegexEngine {

    public static boolean compareChar(char regex, char ch) {
        return regex == ch || regex == '.' || regex == '\0';
    }

    public static boolean compareEqStr(String regex, String str) {
        if (regex.isEmpty() || (regex.equals("$") && str.isEmpty())) {
            return true;
        } else if (regex.length() > 1 && regex.charAt(0) == '\\' && !str.isEmpty()) {
            return compareEqStr(regex.substring(2), str.substring(1));
        } else {
            String regex1 = regex.charAt(0) + regex.substring(2);
            if (regex.length() > 1 && regex.charAt(1) == '?' && !str.isEmpty()) {
                return compareEqStr(regex.substring(2), str) || compareEqStr(regex1, str);
            } else if (regex.length() > 1 && regex.charAt(1) == '*' && !str.isEmpty()) {
                return compareEqStr(regex.substring(2), str) || compareEqStr(regex, str.substring(1));
            } else if (regex.length() > 1 && regex.charAt(1) == '+' && !str.isEmpty()) {
                return compareEqStr(regex1, str) || compareEqStr(regex, str.substring(1));
            } else if (str.isEmpty() || !compareChar(regex.charAt(0), str.charAt(0))) {
                return false;
            } else {
                return compareEqStr(regex.substring(1), str.substring(1));
            }
        }
    }

    public static boolean compareStr(String regex, String str) {
        if (str.isEmpty() && !regex.isEmpty()) {
            return false;
        } else if (compareEqStr(regex, str)) {
            return true;
        } else if (regex.charAt(0) == '^') {
            return compareEqStr(regex.substring(1), str);
        } else {
            return compareStr(regex, str.substring(1));
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String[] inputs = scanner.nextLine().split("\\|");
        String regex = inputs[0];
        String str = inputs[1];
        System.out.println(compareStr(regex, str));
    }
}

File: regex_engine.go
package main

import (
	"fmt"
	"strings"
)

func compareChar(regex byte, ch byte) bool {
	return regex == ch || regex == '.' || regex == 0
}

func compareEqStr(regex string, str string) bool {
	if regex == "" || (regex == "$" && str == "") {
		return true
	} else if len(regex) > 1 && regex[0] == '\\' && str != "" {
		return compareEqStr(regex[2:], str[1:])
	} else if len(regex) > 1 && regex[1] == '?' && str != "" {
		return compareEqStr(regex[2:], str) || compareEqStr(string(regex[0])+regex[2:], str)
	} else if len(regex) > 1 && regex[1] == '*' && str != "" {
		return compareEqStr(regex[2:], str) || compareEqStr(regex, str[1:])
	} else if len(regex) > 1 && regex[1] == '+' && str != "" {
		return compareEqStr(string(regex[0])+regex[2:], str) || compareEqStr(regex, str[1:])
	} else if str == "" || !compareChar(regex[0], str[0]) {
		return false
	} else {
		return compareEqStr(regex[1:], str[1:])
	}
}

func compareStr(regex string, str string) bool {
	if str == "" && regex != "" {
		return false
	} else if compareEqStr(regex, str) {
		return true
	} else if regex[0] == '^' {
		return compareEqStr(regex[1:], str)
	} else {
		return compareStr(regex, str[1:])
	}
}

func main() {
	var regex, str string
	fmt.Scanln(&regex, &str)
	inputs := strings.Split(regex+"|"+str, "|")
	regex = inputs[0]
	str = inputs[1]
	fmt.Println(compareStr(regex, str))
}
